lms/views/dashboard/api/grading.py

```diff
@@ -44,43 +42,50 @@ def __init__(self, request) -> None:
         schema=AutoGradeSyncSchema,
     )
-    def auto_grading_sync(self):
+    def post_auto_grading_sync(self):
         assignment = self.dashboard_service.get_request_assignment(self.request)
-        assert assignment.lis_outcome_service_url, "Assignment without grading URL"
-        lti_registration = self.db.scalars(
-            select(LTIRegistration)
-            .join(ApplicationInstance)
-            .join(Grouping)
-            .where(Grouping.id == assignment.course_id)
-            .order_by(LTIRegistration.updated.desc())
-        ).first()
-        assert lti_registration, "No LTI registraion for LTI1.3 assignment"
 
-        sync_h_user_ids = [g["h_userid"] for g in self.request.parsed_params["grades"]]
+        if self.auto_grading_service.get_in_progress_sync(assignment):
+            self.request.response.status_int = 400
+            return {"message": "There's already a auto-grade sync in progress"}
 
-        sync_lms_users = self.db.execute(
-            select(LMSUser.h_userid, LMSUser.lti_v13_user_id).where(
-                LMSUser.h_userid.in_(sync_h_user_ids)
-            )
+        sync_h_user_ids = [g["h_userid"] for g in self.request.parsed_params["grades"]]
+        sync_lms_users = self.db.scalars(
+            select(LMSUser).where(LMSUser.h_userid.in_(sync_h_user_ids))
         ).all()
-        # Organize the data in a dict h_userid -> lti_user_id for easier access
-        sync_lms_users_by_h_userid = {r[0]: r[1] for r in sync_lms_users}
-
-        grading_service = LTI13GradingService(
-            ltia_service=self.request.find_service(LTIAHTTPService),
-            line_item_url=None,
-            line_item_container_url=None,
-            product_family=None,  # type: ignore
-            misc_plugin=None,  # type: ignore
-            lti_registration=None,  # type: ignore
+        sync_lms_users_by_h_userid: dict[str, LMSUser] = {
+            lms_user.h_userid: lms_user for lms_user in sync_lms_users
+        }
+
+        lms_user_grades = {
+            sync_lms_users_by_h_userid[g["h_userid"]]: g["grade"]
+            for g in self.request.parsed_params["grades"]
+            if g["h_userid"] in sync_lms_users_by_h_userid
+        }
+        grading_sync = self.auto_grading_service.create_grade_sync(
+            assignment,
+            self.request.user.lms_user,
+            lms_user_grades,
         )
-        # Use the same timestamp for all grades of the same sync
-        grade_sync_time_stamp = datetime.now()
-        for grade in self.request.parsed_params["grades"]:
-            grading_service.sync_grade(
-                lti_registration,
-                assignment.lis_outcome_service_url,
-                grade_sync_time_stamp,
-                sync_lms_users_by_h_userid[grade["h_userid"]],
-                grade["grade"],
-            )
+        self.request.add_finished_callback(self._start_sync_grades)
+        return {"id": grading_sync.id, "status": grading_sync.status}
```

Comment:
Do we need to expose the grade ID itself? If every assignment should only have one sync in progress at a time, maybe we could just identify an in-progress or latest sync by the assignment.

That way we don't expose information which could lead to confusions, or change in subsequent calls to the same endpoint. So I would just expose the status for now, and we can evolve from there as we find new needs.

Reply:
Yeah, there's no use for the ID for now, removed it.